Add overrides for `fma` and friends

Clang 14 and later enable [`-ffp-contract=on`](https://releases.llvm.org/14.0.0/tools/clang/docs/UsersManual.html#cmdoption-ffp-contract) by default, which which cause combinations of multiplications/additions to be optimized into the [`llvm.fmuladd.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#llvm-fmuladd-intrinsic) intrinsic, even with `-O0`. Here is an example of a simple program that works when run with `crux-llvm` and Clang 13 and earlier, but fails with Clang 14+ because of this issue:

```c
double fmuladd(double a, double b, double c) {
  return a * b + c;
}

int main(void) {
  double d = fmuladd(1.0, 2.0, 3.0);
  return 0;
}
```
<details>

```
# With Clang 13.0.0
$ cabal run exe:crux-llvm -- test.c -O0
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] All goals discharged through internal simplification.
[Crux] Overall status: Valid.

# With Clang 14.0.0
$ cabal run exe:crux-llvm -- test.c -O0
[Crux] Using pointer width: 64 for file crux-build/crux~test.bc
[Crux] Simulating function main
[Crux] Attempting to prove verification conditions.
[Crux] *** debug executable: results/test/debug-2
[Crux] *** break on line: 2
[Crux] Found counterexample for verification goal
[Crux]   test.c:2:16: error: in fmuladd
[Crux]   Failed to load function handle
[Crux]   Details:
[Crux]     No implementation or override found for pointer
[Crux]     The given pointer could not be resolved to a callable function
[Crux]     No implementation or override found for pointer
[Crux]     Attempting to load callable function with type: double(double, double, double)
[Crux]       Via pointer: Global symbol "llvm.fmuladd.f64" (2, 0x0:[64])
[Crux]     In memory state:
[Crux]       Stack frame fmuladd
[Crux]         Allocations:
[Crux]           StackAlloc 9 0x8:[64] Mutable 8-byte-aligned test.c:1:43
[Crux]           StackAlloc 8 0x8:[64] Mutable 8-byte-aligned test.c:1:33
[Crux]           StackAlloc 7 0x8:[64] Mutable 8-byte-aligned test.c:1:23
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             7 |->   *(7, 0x0:[64]) := 1
[Crux]             8 |->   *(8, 0x0:[64]) := 2
[Crux]             9 |->   *(9, 0x0:[64]) := 3
[Crux]       Stack frame main
[Crux]         Allocations:
[Crux]           StackAlloc 6 0x8:[64] Mutable 8-byte-aligned test.c:6:10
[Crux]           StackAlloc 5 0x4:[64] Mutable 4-byte-aligned internal
[Crux]         Writes:
[Crux]           Indexed chunk:
[Crux]             5 |->   memset (5, 0x0:[64]) 0x0:[8] 0x4:[64]
[Crux]       Base memory
[Crux]         Allocations:
[Crux]           GlobalAlloc 4 0x0:[64] Immutable 1-byte-aligned [defined function ] main
[Crux]           GlobalAlloc 3 0x0:[64] Immutable 1-byte-aligned [defined function ] fmuladd
[Crux]           GlobalAlloc 2 0x0:[64] Immutable 1-byte-aligned [external function] llvm.fmuladd.f64
[Crux]           GlobalAlloc 1 0x0:[64] Immutable 1-byte-aligned [external function] llvm.dbg.declare
[Crux]     in context:
[Crux]       fmuladd
[Crux]       main
[Crux] Goal status:
[Crux]   Total: 1
[Crux]   Proved: 0
[Crux]   Disproved: 1
[Crux]   Incomplete: 0
[Crux]   Unknown: 0
[Crux] Overall status: Invalid.
```
</details>

We should add overrides for `llvm.fmuladd.*`, the related intrinsic [`llvm.fma.*`](https://releases.llvm.org/14.0.0/docs/LangRef.html#int-fma), and the `fma`/`fmaf` functions from `libc`.

--- a/crucible-llvm/doc/limitations.md
+++ b/crucible-llvm/doc/limitations.md
@@ -56,6 +56,14 @@ that arise from invoking certain floating-point operations. For instance,
 invoking the `sqrt` function on a negative value will never result in an
 invalid floating-point exception being raised.
 
+## Floating-point rounding modes
+
+`crucible-llvm` currently makes no effort to track different floating-point
+rounding modes, and the `fegetround`/`fesetround` functions are not yet
+modeled. Whenever the result of a floating-point operation cannot be exactly
+represented in the significand, `crucible-llvm` will default to rounding the
+result toward the nearest even number.
+
 ## Evaluation of floating-point functions
 
 `crucible-llvm` treats the following floating-point operations over `double`s

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics.hs
@@ -283,6 +283,10 @@ declare_overrides =
   , basic_llvm_override LLVM.llvmLog2Override_F64
   , basic_llvm_override LLVM.llvmLog10Override_F32
   , basic_llvm_override LLVM.llvmLog10Override_F64
+  , basic_llvm_override LLVM.llvmFmaOverride_F32
+  , basic_llvm_override LLVM.llvmFmaOverride_F64
+  , basic_llvm_override LLVM.llvmFmuladdOverride_F32
+  , basic_llvm_override LLVM.llvmFmuladdOverride_F64
   , basic_llvm_override LLVM.llvmIsFpclassOverride_F32
   , basic_llvm_override LLVM.llvmIsFpclassOverride_F64
 
@@ -319,6 +323,8 @@ declare_overrides =
   , basic_llvm_override Libc.llvmCeilfOverride
   , basic_llvm_override Libc.llvmFloorOverride
   , basic_llvm_override Libc.llvmFloorfOverride
+  , basic_llvm_override Libc.llvmFmaOverride
+  , basic_llvm_override Libc.llvmFmafOverride
   , basic_llvm_override Libc.llvmIsinfOverride
   , basic_llvm_override Libc.llvm__isinfOverride
   , basic_llvm_override Libc.llvm__isinffOverride

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/LLVM.hs
@@ -1013,6 +1013,54 @@ llvmIsFpclassOverride_F64 =
   [llvmOvr| i1 @llvm.is.fpclass.f64( double, i32 ) |]
   (\_memOps bak args -> Ctx.uncurryAssignment (callIsFpclass bak) args)
 
+llvmFmaOverride_F32 ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat)
+        (FloatType SingleFloat)
+llvmFmaOverride_F32 =
+  [llvmOvr| float @llvm.fma.f32( float, float, float ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (Libc.callFMA bak) args)
+
+llvmFmaOverride_F64 ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat)
+        (FloatType DoubleFloat)
+llvmFmaOverride_F64 =
+  [llvmOvr| double @llvm.fma.f64( double, double, double ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (Libc.callFMA bak) args)
+
+llvmFmuladdOverride_F32 ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat)
+        (FloatType SingleFloat)
+llvmFmuladdOverride_F32 =
+  [llvmOvr| float @llvm.fmuladd.f32( float, float, float ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (Libc.callFMA bak) args)
+
+llvmFmuladdOverride_F64 ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat)
+        (FloatType DoubleFloat)
+llvmFmuladdOverride_F64 =
+  [llvmOvr| double @llvm.fmuladd.f64( double, double, double ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (Libc.callFMA bak) args)
+
 
 llvmX86_pclmulqdq
 --declare <2 x i64> @llvm.x86.pclmulqdq(<2 x i64>, <2 x i64>, i8) #1

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs
@@ -738,6 +738,30 @@ llvmFloorfOverride =
   [llvmOvr| float @floorf( float ) |]
   (\_memOps sym args -> Ctx.uncurryAssignment (callFloor sym) args)
 
+llvmFmafOverride ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat
+                  ::> FloatType SingleFloat)
+        (FloatType SingleFloat)
+llvmFmafOverride =
+  [llvmOvr| float @fmaf( float, float, float ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (callFMA bak) args)
+
+llvmFmaOverride ::
+     forall sym p
+   . IsSymInterface sym
+  => LLVMOverride p sym
+        (EmptyCtx ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat
+                  ::> FloatType DoubleFloat)
+        (FloatType DoubleFloat)
+llvmFmaOverride =
+  [llvmOvr| double @fma( double, double, double ) |]
+  (\_memOps bak args -> Ctx.uncurryAssignment (callFMA bak) args)
+
 
 -- math.h defines isinf() and isnan() as macros, so you might think it unusual
 -- to provide function overrides for them. However, if you write, say,
@@ -869,6 +893,18 @@ callFloor ::
   RegEntry sym (FloatType fi) ->
   OverrideSim p sym ext r args ret (RegValue sym (FloatType fi))
 callFloor bak (regValue -> x) = liftIO $ iFloatRound @_ @fi (backendGetSym bak) RTN x
+
+-- | An implementation of @libc@'s @fma@ function.
+callFMA ::
+     forall fi p sym bak ext r args ret
+   . IsSymBackend sym bak
+  => bak
+  -> RegEntry sym (FloatType fi)
+  -> RegEntry sym (FloatType fi)
+  -> RegEntry sym (FloatType fi)
+  -> OverrideSim p sym ext r args ret (RegValue sym (FloatType fi))
+callFMA bak (regValue -> x) (regValue -> y) (regValue -> z) = liftIO $
+  iFloatFMA @_ @fi (backendGetSym bak) RNE x y z
 
 -- | An implementation of @libc@'s @isinf@ macro. This returns @1@ when the
 -- argument is positive infinity, @-1@ when the argument is negative infinity,

--- a/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs
+++ b/crucible-llvm/src/Lang/Crucible/LLVM/Intrinsics/Libc.hs
@@ -904,7 +904,7 @@ callFMA ::
   -> RegEntry sym (FloatType fi)
   -> OverrideSim p sym ext r args ret (RegValue sym (FloatType fi))
 callFMA bak (regValue -> x) (regValue -> y) (regValue -> z) = liftIO $
-  iFloatFMA @_ @fi (backendGetSym bak) RNE x y z
+  iFloatFMA @_ @fi (backendGetSym bak) defaultRM x y z
 
 -- | An implementation of @libc@'s @isinf@ macro. This returns @1@ when the
 -- argument is positive infinity, @-1@ when the argument is negative infinity,
@@ -951,7 +951,7 @@ callSqrt ::
   bak ->
   RegEntry sym (FloatType fi) ->
   OverrideSim p sym ext r args ret (RegValue sym (FloatType fi))
-callSqrt bak (regValue -> x) = liftIO $ iFloatSqrt @_ @fi (backendGetSym bak) RNE x
+callSqrt bak (regValue -> x) = liftIO $ iFloatSqrt @_ @fi (backendGetSym bak) defaultRM x
 
 ------------------------------------------------------------------------
 -- **** Circular trigonometry functions
@@ -1696,3 +1696,12 @@ cxa_atexitOverride
 cxa_atexitOverride =
   [llvmOvr| i32 @__cxa_atexit( void (i8*)*, i8*, i8* ) |]
   (\_ bak _args -> liftIO $ bvLit (backendGetSym bak) knownNat (BV.zero knownNat))
+
+----------------------------------------------------------------------------
+
+-- | IEEE 754 declares 'RNE' to be the default rounding mode, and most @libc@
+-- implementations agree with this in practice. The only places where we do not
+-- use this as the default are operations that specifically require the behavior
+-- of a particular rounding mode, such as @ceil@ or @floor@.
+defaultRM :: RoundingMode
+defaultRM = RNE

--- a/crux-llvm/test-data/golden/fmuladd.c
+++ b/crux-llvm/test-data/golden/fmuladd.c
@@ -1,0 +1,27 @@
+// A regression test for https://github.com/GaloisInc/crucible/issues/1154.
+#include <crucible.h>
+#include <math.h>
+
+// LLVM is liable to optimize these functions in a way that uses the
+// llvm.fmuladd.* or llvm.fma.* intrinsics.
+float fmuladdf(float a, float b, float c) {
+  return a * b + c;
+}
+
+double fmuladd(double a, double b, double c) {
+  return a * b + c;
+}
+
+int main(void) {
+  float f1 = fmuladdf(1.0f, 2.0f, 3.0f);
+  check(f1 == f1);
+  float f2 = fmaf(1.0f, 2.0f, 3.0f);
+  check(f2 == f2);
+
+  double d1 = fmuladd(1.0, 2.0, 3.0);
+  check(d1 == d1);
+  double d2 = fma(1.0, 2.0, 3.0);
+  check(d2 == d2);
+
+  return 0;
+}

--- a/crux-llvm/test-data/golden/fmuladd.config
+++ b/crux-llvm/test-data/golden/fmuladd.config
@@ -1,0 +1,1 @@
+opt-level: 0

--- a/crux-llvm/test-data/golden/fmuladd.z3.good
+++ b/crux-llvm/test-data/golden/fmuladd.z3.good
@@ -1,0 +1,1 @@
+[Crux] Overall status: Valid.

